pkg/autoupdate/tuf/library_manager.go

```diff
@@ -0,0 +1,341 @@
+package tuf
+
+import (
+	"context"
+	"crypto/sha512"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"path/filepath"
+	"sort"
+	"strings"
+
+	"github.com/Masterminds/semver"
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/kolide/kit/fsutil"
+	"github.com/kolide/kit/version"
+	"github.com/kolide/launcher/ee/localserver"
+	"github.com/kolide/launcher/pkg/autoupdate"
+	client "github.com/theupdateframework/go-tuf/client"
+)
+
+// updateLibraryManager manages the update libraries for launcher and osquery.
+// It downloads and verifies new updates, and moves them to the appropriate
+// location in the library specified by the version associated with that update.
+// It also ensures that old updates are removed when they are no longer needed.
+type updateLibraryManager struct {
+	metadataClient  *client.Client // used to validate downloads
+	mirrorUrl       string         // dl.kolide.co
+	mirrorClient    *http.Client
+	rootDirectory   string
+	operatingSystem string
+	osquerier       localserver.Querier // used to query for current running osquery version
+	logger          log.Logger
+}
+
+func newUpdateLibraryManager(metadataClient *client.Client, mirrorUrl string, mirrorClient *http.Client, rootDirectory string, operatingSystem string, osquerier localserver.Querier, logger log.Logger) (*updateLibraryManager, error) {
+	ulm := updateLibraryManager{
+		metadataClient:  metadataClient,
+		mirrorUrl:       mirrorUrl,
+		mirrorClient:    mirrorClient,
+		rootDirectory:   rootDirectory,
+		operatingSystem: operatingSystem,
+		osquerier:       osquerier,
+		logger:          log.With(logger, "component", "tuf_autoupdater_library_manager"),
+	}
+
+	// Ensure our staged updates and updates directories exist
+	for _, binary := range binaries {
+		// Attempt out the staged updates directory, in case there are any archives there that we
+		// were unable to clean up previously.
+		_ = os.RemoveAll(ulm.stagedUpdatesDirectory(binary))
+
+		// Create the directory for staging updates
+		if err := os.MkdirAll(ulm.stagedUpdatesDirectory(binary), 0755); err != nil {
+			return nil, fmt.Errorf("could not make staged updates directory for %s: %w", binary, err)
+		}
+
+		// Create the update library
+		if err := os.MkdirAll(ulm.updatesDirectory(binary), 0755); err != nil {
+			return nil, fmt.Errorf("could not make updates directory for %s: %w", binary, err)
+		}
+	}
+
+	return &ulm, nil
+}
+
+// updatesDirectory returns the update library location for the given binary.
+func (ulm *updateLibraryManager) updatesDirectory(binary string) string {
+	return filepath.Join(ulm.rootDirectory, fmt.Sprintf("%s-updates", binary))
+}
+
+// stagedUpdatesDirectory returns the location for staged updates -- i.e. updates
+// that have been downloaded, but not yet verified.
+func (ulm *updateLibraryManager) stagedUpdatesDirectory(binary string) string {
```

Comment:
What happens if, somehow, an update triggers while one is running? They'll both end up in the same staging directory.

Reply:
I didn't put it in this PR because it felt like it already contained enough changes, but my plan was to add two workers and a wait group to the TUF autoupdater, so that the `addToLibrary` calls for both launcher + osqueryd could run at the same time without one cutting off the other and triggering a launcher restart too soon. I think that'd also prevent an update happening when one is already running. Let me know what you think or if you have another suggestion?

Reply:
Sure, that's a good point. I'll update.

Reply:
Added to the other PR here 🙂 https://github.com/kolide/launcher/pull/1119/commits/c75e4385550bab643665b788e38c94433a54608c